crates/pera-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,483 @@
+            .ok_or(DwalletMPCError::EpochEnded(self.epoch_id))
+    }
+
+    /// Handles a message by forwarding it to the relevant MPC session.
+    /// If the session does not exist, punish the sender.
+    pub(crate) fn handle_message(
+        &mut self,
+        message: &[u8],
+        authority_name: AuthorityName,
+        session_id: ObjectID,
+    ) -> DwalletMPCResult<()> {
+        if self.malicious_actors.contains(&authority_name) {
+            return Ok(());
+        }
+        let session = match self.mpc_sessions.get_mut(&session_id) {
+            Some(session) => session,
+            None => {
+                warn!(
+                    "received a message for an MPC session ID: `{:?}` which does not exist",
+                    session_id
+                );
+                self.malicious_actors.insert(authority_name);
+                return Ok(());
+            }
+        };
+        match session.handle_message(&DWalletMPCMessage {
+            message: message.to_vec(),
+            authority: authority_name,
+        }) {
+            Err(DwalletMPCError::MaliciousParties(malicious_parties)) => {
+                self.flag_parties_as_malicious(&malicious_parties)?;
+                Ok(())
+            }
+            other => other,
+        }
+    }
+
+    /// Convert the indices of the malicious parties to their addresses and store them
+    /// in the malicious actors set.
+    /// New messages from these parties will be ignored.
+    /// todo(zeev): clarify if it's restarted on epoch change.
+    fn flag_parties_as_malicious(&mut self, malicious_parties: &[PartyID]) -> DwalletMPCResult<()> {
+        let malicious_parties_names = malicious_parties
+            .iter()
+            .map(|party_id| {
+                self.epoch_store()?
+                    .committee()
+                    .authority_by_index(*party_id as u32)
+                    .cloned()
+                    .ok_or(DwalletMPCError::AuthorityIndexNotFound(*party_id))
+            })
+            .collect::<DwalletMPCResult<Vec<AuthorityName>>>()?;
+        warn!(
+            "[dWallet MPC] Flagged the following parties as malicious: {:?}",
+            malicious_parties_names
+        );
+        self.malicious_actors.extend(malicious_parties_names);
+        Ok(())
+    }
+
+    /// Spawns a new MPC session if the number of active sessions is below the limit.
+    /// Otherwise, add the session to the pending queue.
+    pub(crate) fn push_new_mpc_session(
+        &mut self,
+        auxiliary_input: Vec<u8>,
+        party: MPCParty,
+        session_info: SessionInfo,
+    ) -> DwalletMPCResult<()> {
+        if self.mpc_sessions.contains_key(&session_info.session_id) {
+            // This should never happen, as the session ID is a Move UniqueID.
+            error!(
+                "received start flow event for session ID {:?} that already exists",
+                &session_info.session_id
+            );
+            return Ok(());
```

Comment:
I'm not sure returning `Ok(())` is the way to go here. This error should never occur, true. However, if it occurrs and you simply continue, what will happen? Potentially, there would be a mix-up between messages of different rounds of the old session and new one. 

Another thing is: the session id is actually an object ID. In the case of presign for example, two sign sessions must never use the same presign, so if we encounter this error it might mean some kind of attack is undergoing when it shouldn't.

Let's think carefully through implications. Perhaps the best thing to do here is kill the ongoing session and don't open a new one.